crates/runtime/src/lib.rs

```diff
@@ -142,6 +145,29 @@ impl<'a, T> GuestPtr<'a, [T]> {
         let base = GuestPtr::new(self.mem, self.offset_base());
         (0..self.len()).map(move |i| base.add(i))
     }
+
+    /// Safety note: The user must use the same `GuestBorrows` to check each
+    /// use of GuestType::as_raw that is alive simultaneously.
+    /// The resulting `*mut [T]` can be used as a `&mut [T]` as long as the
+    /// reference is dropped before any Wasm code is re-entered.
+    pub fn as_raw(&self, bc: &mut GuestBorrows) -> Result<*mut [T], GuestError> {
```

Comment:
One of the reasons I hesitated about returning `*mut [T]` is that given the design of `GuestType` we can't actually do that. One of the purposes of `GuestType` is to have a layer of indirection which can allow for different guest/host representations (e.g. a guest pointer is `u32` but on the host it's a `u32` plus the guest memory object) and also validation (e.g. enums only have some valid bit patterns).

Now I think it's still fine to have a method like this, but we would likely need to sternly document that an additional safety guarantee for using the returned pointer is that `T` must be some sort of primitive type like `u8` or such.

From an API perspective I'm thinking that `GuestPtr` is pretty low-level so we probably don't want to have the `GuestBorrows` type here, but instead `GuestBorrows` could have like an `get_raw` method which is a bit safer than this method. I tried do add documentation to `GuestMemory` (which this would reference) about aliasing, and we should also expand that with how to use `GuestBorrows` to manage the dynamic runtime aliasing.

Reply:
That additional safety guarantee about T is important, and something I overlooked. Thats linked to what I said above about `GuestType::validate`, which used to take care of that guarantee.

Reply:
I also support moving these to be methods of `GuestBorrows`, rather than methods on pointers.

And, last of all, I'm open to a better name than `GuestBorrows` if anyone thinks of one.

Reply:
Oh and to be clear I think it's fine to have this method here, for example it's mega-useful for `GuestPtr<[u8]>` which is one of the main reasons for its existence.

I think it's fine to say in the documentation that there's just more guarantees you need to adhere to in order to safely use the raw pointer being returned.

Reply:
I ended up having to re-introduce a GuestTypeTransparent to express the kinds of types its OK to use a ref in linear memory to. This is all types not containing pointers or unions.

I think some unions may actually be OK to be transparent but I don't know how Rust lays out the discriminant in enums. @sw17ch recently had some fun with this but maybe you know as well.